File: .claude/skills/agent-skill-kit-commons/packages/py/src/skills_ref/audit/path_leak.py
from __future__ import annotations

import re
from pathlib import Path

_HOST_PATH_RE = re.compile(r"(/Users/|/home/|/tmp/)")
_TEXT_EXT = frozenset(
    {".sh", ".bash", ".py", ".mjs", ".js", ".ts", ".md", ".yaml", ".yml", ".json", ".toml", ".txt", ""}
)


def _walk_all(d: Path) -> list:
    out = []
    try:
        entries = sorted(d.iterdir(), key=lambda p: p.name)
    except OSError:
        return out
    for entry in entries:
        if entry.is_dir():
            out.extend(_walk_all(entry))
        elif entry.is_file():
            out.append(entry)
    return out


_INLINE_BACKTICK_RE = re.compile(r"`[^`]*`")


def _body_scan_lines(body: str):
    lines = body.split("\n")
    in_fence = False
    skip_fence = False
    out = []
    for i, line in enumerate(lines):
        m = re.match(r"^```(\S*)", line)
        if m:
            if not in_fence:
                in_fence = True
                lang = (m.group(1) or "").lower()
                skip_fence = lang in ("text", "console", "bash")
            else:
                in_fence = False
                skip_fence = False
            continue
        if in_fence and skip_fence:
            continue
        if line.lstrip().startswith(">"):
            continue
        scrubbed = _INLINE_BACKTICK_RE.sub("", line)
        out.append((scrubbed, i, line))
    return out


def audit_path_leak(registry) -> list:
    findings = []
    for name, node in registry.nodes.items():
        d = node.get("dir")
        if not d:
            continue
        body = node.get("body") or ""
        if body:
            for scrubbed, i, raw in _body_scan_lines(body):
                if _HOST_PATH_RE.search(scrubbed):
                    findings.append(
                        {
                            "severity": "MEDIUM",
                            "code": "A030_HOST_PATH_LEAK",
                            "where": f"{d}/SKILL.md:body+{i + 1}",
                            "message": f"host-path literal in body: {raw.strip()}",
                        }
                    )
        scripts_dir = Path(d) / "scripts"
        if not scripts_dir.is_dir():
            continue
        for file in _walk_all(scripts_dir):
            if file.suffix not in _TEXT_EXT:
                continue
            try:
                text = file.read_text(encoding="utf-8")
            except OSError:
                continue
            for i, line in enumerate(text.split("\n")):
                if _HOST_PATH_RE.search(line):
                    findings.append(
                        {
                            "severity": "MEDIUM",
                            "code": "A030_HOST_PATH_LEAK",
                            "where": f"{file}:{i + 1}",
                            "message": f"host-path literal: {line.strip()}",
                        }
                    )
    return findings

File: .claude/skills/agent-skill-kit-commons/packages/mjs/src/audit/path_leak.mjs
import { readFileSync, readdirSync, statSync } from "node:fs";
import { join, extname } from "node:path";

const HOST_PATH_RE = /(\/Users\/|\/home\/|\/tmp\/)/;

function walkAll(dir) {
  const out = [];
  let entries;
  try {
    entries = readdirSync(dir, { withFileTypes: true });
  } catch {
    return out;
  }
  entries.sort((a, b) => (a.name < b.name ? -1 : a.name > b.name ? 1 : 0));
  for (const entry of entries) {
    const p = join(dir, entry.name);
    if (entry.isDirectory()) out.push(...walkAll(p));
    else if (entry.isFile()) out.push(p);
  }
  return out;
}

/**
 * Strip fenced code blocks whose language tag is text/console/bash from the body
 * before scanning. Markdown blockquote lines (starting with '>') are skipped.
 * Inline backtick spans are blanked out so example paths shown as `code` don't
 * trip the host-path regex.
 */
function bodyScanLines(body) {
  const lines = body.split("\n");
  const out = [];
  let inFence = false;
  let skipFence = false;
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    const fenceMatch = line.match(/^```(\S*)/);
    if (fenceMatch) {
      if (!inFence) {
        inFence = true;
        const lang = (fenceMatch[1] ?? "").toLowerCase();
        skipFence = lang === "text" || lang === "console" || lang === "bash";
      } else {
        inFence = false;
        skipFence = false;
      }
      continue;
    }
    if (inFence && skipFence) continue;
    if (line.trimStart().startsWith(">")) continue;
    // Blank out inline backtick spans so example paths in docs don't trip.
    const scrubbed = line.replace(/`[^`]*`/g, "");
    out.push({ line: scrubbed, idx: i, raw: line });
  }
  return out;
}

export function auditPathLeak(registry) {
  const findings = [];
  for (const [, node] of registry.nodes) {
    if (!node.dir) continue;
    // Body scan: strip carve-out blocks.
    const body = node.body ?? "";
    if (body) {
      for (const { line, idx, raw } of bodyScanLines(body)) {
        if (HOST_PATH_RE.test(line)) {
          findings.push({
            severity: "MEDIUM",
            code: "A030_HOST_PATH_LEAK",
            where: `${node.dir}/SKILL.md:body+${idx + 1}`,
            message: `host-path literal in body: ${(raw ?? line).trim()}`,
          });
        }
      }
    }
    // Scripts scan: all text files under scripts/.
    const scriptsDir = join(node.dir, "scripts");
    try {
      if (!statSync(scriptsDir).isDirectory()) continue;
    } catch {
      continue;
    }
    for (const file of walkAll(scriptsDir)) {
      // Only scan text-ish files: extension list is best-effort.
      const ext = extname(file);
      const TEXT_EXT = new Set([
        ".sh",
        ".bash",
        ".py",
        ".mjs",
        ".js",
        ".ts",
        ".md",
        ".yaml",
        ".yml",
        ".json",
        ".toml",
        ".txt",
        "",
      ]);
      if (!TEXT_EXT.has(ext)) continue;
      let text;
      try {
        text = readFileSync(file, "utf8");
      } catch {
        continue;
      }
      const lines = text.split("\n");
      for (let i = 0; i < lines.length; i++) {
        if (HOST_PATH_RE.test(lines[i])) {
          findings.push({
            severity: "MEDIUM",
            code: "A030_HOST_PATH_LEAK",
            where: `${file}:${i + 1}`,
            message: `host-path literal: ${lines[i].trim()}`,
          });
        }
      }
    }
  }
  return findings;
}
